hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/async/RouterAsyncClientProtocol.java

```diff
@@ -0,0 +1,1083 @@
+        replication, blockSize, supportedVersions, ecPolicyName, storagePolicy);
+    final List<RemoteLocation> locations =
+        rpcServer.getLocationsForPath(src, true);
+    final RemoteLocation[] createLocation = new RemoteLocation[1];
+    asyncTry(() -> {
+      rpcServer.getCreateLocationAsync(src, locations);
+      asyncApply((AsyncApplyFunction<RemoteLocation, Object>) remoteLocation -> {
+        createLocation[0] = remoteLocation;
+        rpcClient.invokeSingle(remoteLocation, method, HdfsFileStatus.class);
+        asyncApply((ApplyFunction<HdfsFileStatus, Object>) status -> {
+          status.setNamespace(remoteLocation.getNameserviceId());
+          return status;
+        });
+      });
+    });
+    asyncCatch((AsyncCatchFunction<Object, IOException>) (o, ioe) -> {
+      final List<RemoteLocation> newLocations = checkFaultTolerantRetry(
+          method, src, ioe, createLocation[0], locations);
+      rpcClient.invokeSequential(
+          newLocations, method, HdfsFileStatus.class, null);
+    }, IOException.class);
+
+    return asyncReturn(HdfsFileStatus.class);
+  }
+
+  @Override
+  public LastBlockWithStatus append(
+      String src, String clientName,
+      EnumSetWritable<CreateFlag> flag) throws IOException {
+    rpcServer.checkOperation(NameNode.OperationCategory.WRITE);
+
+    List<RemoteLocation> locations = rpcServer.getLocationsForPath(src, true);
+    RemoteMethod method = new RemoteMethod("append",
+        new Class<?>[] {String.class, String.class, EnumSetWritable.class},
+        new RemoteParam(), clientName, flag);
+    rpcClient.invokeSequential(method, locations, LastBlockWithStatus.class, null);
+    asyncApply((ApplyFunction<RemoteResult, LastBlockWithStatus>) result -> {
+      LastBlockWithStatus lbws = (LastBlockWithStatus) result.getResult();
+      lbws.getFileStatus().setNamespace(result.getLocation().getNameserviceId());
+      return lbws;
+    });
+    return asyncReturn(LastBlockWithStatus.class);
+  }
+
+  @Deprecated
+  @Override
+  public boolean rename(final String src, final String dst)
+      throws IOException {
+    rpcServer.checkOperation(NameNode.OperationCategory.WRITE);
+
+    final List<RemoteLocation> srcLocations =
+        rpcServer.getLocationsForPath(src, true, false);
+    final List<RemoteLocation> dstLocations =
+        rpcServer.getLocationsForPath(dst, false, false);
+    // srcLocations may be trimmed by getRenameDestinations()
+    final List<RemoteLocation> locs = new LinkedList<>(srcLocations);
+    RemoteParam dstParam = getRenameDestinations(locs, dstLocations);
+    if (locs.isEmpty()) {
+      asyncComplete(
+          rbfRename.routerFedRename(src, dst, srcLocations, dstLocations));
+      return asyncReturn(Boolean.class);
+    }
+    RemoteMethod method = new RemoteMethod("rename",
+        new Class<?>[] {String.class, String.class},
+        new RemoteParam(), dstParam);
+    isMultiDestDirectory(src);
```

Comment:
Duplicate name here?

Reply:
Addendum: What this method invoke want to do? Thanks.

Reply:
HI, @Hexiaoqiao, this is the asynchronous implementation corresponding to the rename method of RouterClientProtocol below. It calls isMultiDestDirectory and obtains the return value of isMultiDestDirectory through asyncApply(...) for subsequent processing.
```java
    // RouterClientProtocol#rename
    RemoteMethod method = new RemoteMethod("rename",
        new Class<?>[] {String.class, String.class},
        new RemoteParam(), dstParam);
    if (isMultiDestDirectory(src)) {
      if (locs.size() != srcLocations.size()) {
        throw new IOException("Rename of " + src + " to " + dst + " is not"
            + " allowed. The number of remote locations for both source and"
            + " target should be same.");
      }
      return rpcClient.invokeAll(locs, method);
    } else {
      return rpcClient.invokeSequential(locs, method, Boolean.class,
          Boolean.TRUE);
    }
```

Reply:
Got it, thanks. Others look good to me.